### PR TITLE
Feature/wip2: Adding paging support for datasets endpoint

### PR DESCRIPTION
**Pull request by aashishlalani**
The paginate function was also generalized and a unit test for it was added.

### CODE REVIEW

File: api2-derived-dataset-client/src/main/scala/com/tapad/derived/dataset/apiclient/DerivedDatasetClient.scala [hunk 1]
```diff
@@ -31,12 +32,59 @@ object DerivedDatasetClient {
...
+
+  // retry function
+  def retry[T](block: => Future[T], tries: Int): Future[T] = {
+    block recoverWith {
+      case _ if tries > 0 => retry(block, tries - 1)
+    }
+  }
+
+  // get the total number of pages- up to 5 retries on failure
+  def getFirst(f: Future[DerivedDatasetListDto]) = {
```
Comment (yury-skobov) [inline api2-derived-dataset-client/src/main/scala/com/tapad/derived/dataset/apiclient/DerivedDatasetClient.scala]: Let's make the the number of retries configurable. 5 is a fine default value.

File: api2-derived-dataset-client/src/main/scala/com/tapad/derived/dataset/apiclient/DerivedDatasetClient.scala [hunk 2]
```diff
@@ -31,12 +32,59 @@ object DerivedDatasetClient {
...
+    val seqDto = retry(f, 5)
+    seqDto onComplete {
+      case Failure(e) => println(e)
```
Comment (yury-skobov) [inline api2-derived-dataset-client/src/main/scala/com/tapad/derived/dataset/apiclient/DerivedDatasetClient.scala]: Log, not println.

Comment (yury-skobov) [reply api2-derived-dataset-client/src/main/scala/com/tapad/derived/dataset/apiclient/DerivedDatasetClient.scala]: Also, if you're only interested in the failure case, there is an `onFailure`.

File: api2-derived-dataset-client/src/main/scala/com/tapad/derived/dataset/apiclient/DerivedDatasetClient.scala [hunk 3]
```diff
@@ -58,21 +106,30 @@ class DerivedDatasetClient(
...
+    // resolve total number of pages
+    val firstRes = Await.result(getFirst(addParamsAndExecute[DerivedDatasetListDto](requestWithHeaders.addQueryStringParameters, requestWithFilters(0, 1))), 1 minute)
+    val (firstResData, totalCount) = (firstRes.data, firstRes.totalCount)
+
+    paginate[DerivedDatasetListDto, DerivedDatasetDto](addParamsAndExecute[DerivedDatasetListDto], requestWithHeaders.addQueryStringParameters, requestWithFilters, 1, 1, totalCount, firstResData, dto => dto.data)
   }
 
   def retrievedAllPagesInResponse(response: StandaloneWSResponse): Seq[Any] = ???
 }
+
```
Comment (yury-skobov) [inline api2-derived-dataset-client/src/main/scala/com/tapad/derived/dataset/apiclient/DerivedDatasetClient.scala]: Looks like `retrievedAllPagesInResponse` is still missing implementation.

Comment (aashishlalani) [reply api2-derived-dataset-client/src/main/scala/com/tapad/derived/dataset/apiclient/DerivedDatasetClient.scala]: This was actually leftover from Eli's previous commit (just had to fix it up a bit to make it compile)

File: api2-derived-dataset-client/src/main/scala/com/tapad/derived/dataset/apiclient/DerivedDatasetClient.scala [hunk 4]
```diff
@@ -30,6 +30,7 @@ object DerivedDatasetClient {
   case class QueryExistsFilter(filter: String = "query exists") extends Api2Filter
   case class DeleteByNotNullFilter(filter: String = "deletedBy is not null") extends Api2Filter
   case class DeleteByIsNullFilter(filter: String = "deletedBy is null") extends Api2Filter
+  case class DeleteByNullFilter(filter: String = "deletedBy is null") extends Api2Filter
```
Comment (elitaveras) [inline api2-derived-dataset-client/src/main/scala/com/tapad/derived/dataset/apiclient/DerivedDatasetClient.scala]: this is a repeat erase one of the two

**Pull request merged**